### PR TITLE
Add simsiam tutorial

### PR DESCRIPTION
**Pull request by philippmwirth**
# SimSiam Tutorial: Sentinel-2 Dataset Analysis

Made a tutorial using the recently added SimSiam model. The tutorial highlights how to train and embed images with the SimSiam model in *old-school PyTorch* style. 

Furthermore, I fixed a typo in the SimCLR tutorial and added an alternative way to download the docker image when the initial command has failed.

### CODE REVIEW

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+# %%
+# Setup data augmentations and loaders
+# ------------------------------------
+# Since we're working on satellite images, it makes sence to use horizonal and
+# vertical flips as well as random rotation transformations. We apply weak color 
+# jitter to learn an invariance of the model with respect to slight changes in
+# the color of the water.
+#
+
+# the collate function defines the transformations for self-supervised learning
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:90]: The term 'data augmentation' is never used here, even though it is the most common word for such operations (not 'transformations' or even 'collate functions'. This confused me.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+
+# create a torchvision transformation for embedding the dataset after training
+test_transforms = torchvision.transforms.Compose([
+    torchvision.transforms.Resize((input_size, input_size)),
+    torchvision.transforms.ToTensor(),
+    torchvision.transforms.Normalize(
+        mean=lightly.data.collate.imagenet_normalize['mean'],
+        std=lightly.data.collate.imagenet_normalize['std'],
+    )
+])
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:113]: No reasoning is given for these transformations, it might be good to explain them.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+
+# create a lightly dataset for training (no transforms!)
+dataset_train_simsiam = lightly.data.LightlyDataset(
+    input_dir=path_to_data
+)
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:118]: It might be good to explain why no transforms are made here, even though the data augmentation transforms are needed only for training but not for testing. In general, it took me quite a while to understand which of the two transformations defined before is needed for training / testing and why this is so. 
Perhaps it is easier to understand if the order of the 6 blocks is changed: First the collate function (data augmentation function) is defined, then the train dataset, then the train data loader. After that, the three blocks for embedding images are explained: the resizing transform, the dataset definition and the data loader definition.
It should also be made clear why the resizing transform is added to the dataset, but the data augmentation transform to the data loader.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+    momentum=0.9,
+    weight_decay=5e-4
+)
+
+
+# %% 
+# Use a GPU if available
+
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:188]: Perhaps put this block in line 208 just before the 'model.to(device)' as they belong together.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+        # calculate the per-dimension standard deviation of the outputs
+        # we can use this later to check whether the embeddings are collapsing
+        output, _ = y0
+        output = output.detach()
+        output = torch.nn.functional.normalize(output, dim=1)
+        
+        output_std = torch.std(output, 0)
+        output_std = output_std.mean()
+
+        # use running averages to track the loss and standard deviation
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:239]: 'running averages' -> 'moving averages'

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+import umap
+from sklearn import random_projection
+
+# %%
+# Then, we transform the embeddings using UMAP and rescale them to fit in the 
+# [0, 1] square.
+#
+
+
+# use a random Gaussian projection to get 2d embeddings
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:310]: You could explain why we need 2d-embeddings here and thus the projection.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+]
+
+
+def get_image_as_np_array(filename: str):
+    """Returns an image as a numpy array.
+    """
+    img = Image.open(filename)
+    return np.asarray(img)
+
+
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:397]: Some comments are very obvious and essentially say the same as the line of code above or below in the same words. Such comments do not add any value to the reader, but are just another thing to read. This applies to following comments: 

>  392: """Returns an image as a numpy array."""
> 402: " get shape of the image "
> 429: "add the subplot"
> 436: "plot the image"
> 441: "disable the axis"

I would rather have fewer and larger comments explaining the idea/summary of what is done than a comment for every line of code.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+    """Plots the example image and its eight nearest neighbors.
+    """
+    n_subplots = 9
+    # initialize empty figure
+    fig = plt.figure()
+    fig.suptitle(f"Nearest Neighbor Plot {i + 1}")
+    #
+    example_idx = filenames.index(example_image)
+    # get distances to the cluster center
+    distances = embeddings - embeddings[example_idx]
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:422]: Why not simply use np.pairwise_distances() ?

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+    distances = np.power(distances, 2).sum(-1).squeeze()
+    # sort indices by distance to the center
+    argsort = np.argsort(distances)
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:425]: 'argsort' is quite a bad name, because it is also a function. Furthermore doing a complete sort is not efficient, replace it with `indices_closest_images = np.argpartition(distances,n_subplots-1)[:n_subplots]`

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+    # show images
+    plot_indices = argsort[:n_subplots]
+    for plot_offset in range(n_subplots):
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:428]: Why not `for plot_offset, image_id in enumerate(closest_indices):` ? It would also replace line 432

File: docs/source/tutorials_source/package/tutorial_simclr_clothing.py [hunk 11]
```diff
@@ -133,7 +133,9 @@
 # %%
 # Create the SimCLR model
 # -----------------------
```
Comment (MalteEbner) [inline docs/source/tutorials_source/package/tutorial_simclr_clothing.py:135]: That does not belong in this commit

File: docs/source/docker/getting_started/setup.rst [hunk 12]
```diff
@@ -48,6 +48,12 @@ We assume *container-credentials.json* is in your current directory.
 
     cat container-credentials.json | docker login -u json_key --password-stdin https://eu.gcr.io
 
+If the above command does not work, try the following:
```
Comment (MalteEbner) [inline docs/source/docker/getting_started/setup.rst:51]: That does not belong in this commit

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py (hunk 8 above)
Comment (philippmwirth) [reply docs/source/tutorials_source/package/tutorial_simsiam_esa.py:422]: I am not aware of such a Numpy function, do you mean [this one](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise_distances.html)? In any case I think the code here is fine.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py (hunk 9 above)
Comment (philippmwirth) [reply docs/source/tutorials_source/package/tutorial_simsiam_esa.py:425]: The advantage of the complete sort is that the first image is still the example image whereas with argpartition the order of the indices is undefined. Furthermore, this line is by far no bottleneck - I will continue using the full sort but change the name.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py (hunk 8 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/package/tutorial_simsiam_esa.py:423]: You are right, it is from sklearn. I just thought about it, as it is used for the python implementation of coreset.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py (hunk 9 above)
Comment (MalteEbner) [reply docs/source/tutorials_source/package/tutorial_simsiam_esa.py]: If it is not the bottleneck, then the performance is indeed irrelevant. I didn't think about the order within the 9 nearest images being relevant, sorry.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+dataset_train_simsiam = lightly.data.LightlyDataset(
+    input_dir=path_to_data
+)
+
+# create a dataloader for training
+dataloader_train_simsiam = torch.utils.data.DataLoader(
+    dataset_train_simsiam,
+    batch_size=batch_size,
+    shuffle=True,
+    collate_fn=collate_fn,
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:117]: I would add `drop_last=True` for training of the models. Besides performance boost (because the input tensor shape is always the same) it could also help with stability during training.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+# vectors. If it is close to one divided by the square root of the output 
+# dimension, everything is fine (you can read
+# up on this idea `here <https://arxiv.org/abs/2011.10566>`_).
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
+model.to(device)
+
+avg_loss = 0.
+avg_output_std = 0.
+for e in range(1, epochs + 1):
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:213]: Would just do `for epoch in range(epochs):`. Maybe I missed it, but I don't see any problem having the first epoch be 0.

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+        # backpropagation
+        loss = criterion(y0, y1)
+        loss.backward()
+
+        optimizer.step()
+        optimizer.zero_grad()
+
+        # calculate the per-dimension standard deviation of the outputs
+        # we can use this later to check whether the embeddings are collapsing
+        output, _ = y0
```
Comment (IgorSusmelj) [inline docs/source/tutorials_source/package/tutorial_simsiam_esa.py:233]: I miss somehow the information that the SimSiam forward pass returns a tuple of (prediction, projection). Maybe adding a comment or renaming the variables would help?

File: docs/source/tutorials_source/package/tutorial_simsiam_esa.py (hunk 13 above)
Comment (philippmwirth) [reply docs/source/tutorials_source/package/tutorial_simsiam_esa.py:117]: Ok, I don't mind adding `drop_last=True`. Not sure it has such a big impact though.

**Review by IgorSusmelj — APPROVED** [review]
